pkg/tekton/metapipeline/clientfactory.go

```diff
@@ -0,0 +1,280 @@
+package metapipeline
+
+import (
+	"fmt"
+	"github.com/jenkins-x/jx/pkg/apps"
+	"github.com/jenkins-x/jx/pkg/client/clientset/versioned"
+	jxclient "github.com/jenkins-x/jx/pkg/client/clientset/versioned"
+	"github.com/jenkins-x/jx/pkg/config"
+	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/jxfactory"
+	"github.com/jenkins-x/jx/pkg/kube"
+	"github.com/jenkins-x/jx/pkg/log"
+	"github.com/jenkins-x/jx/pkg/prow"
+	"github.com/jenkins-x/jx/pkg/tekton"
+	"github.com/jenkins-x/jx/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	tektonclient "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"io/ioutil"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"os"
+	"time"
+)
+
+const (
+	retryDuration      = time.Second * 30
+	defaultCheckoutDir = "source"
+)
+
+var (
+	logger = log.Logger().WithFields(logrus.Fields{"component": "meta-pipeline-client"})
+)
+
+// clientFactory implements the interface methods to create and apply the meta pipeline.
+type clientFactory struct {
+	Client
```

Comment:
I don't get why the `Client` interface needs to be embedded into `clientFactory` type. The type is implementing already this interface. 

This pattern is a form of composition in go (https://rakyll.org/typesystem/). Normally, it is used to embed methods of other types into a type. 

In this case you are save to remove it.

Reply:
I removed it

Reply:
Prior to when the constructor was returning a concrete implementation instead of the interface adding `Client` would ensure that you implement all methods. But this is moot now anyway since the constructor returns the interface.